controller/api/destination/watcher/k8s.go

```diff
@@ -50,25 +50,41 @@ func getHostAndPort(authority string) (string, Port, error) {
 // GetServiceAndPort is a utility function that destructures an authority into
 // a service and port.  If the authority does not represent a Kubernetes
 // service, an error is returned.  If no port is specified in the authority,
-// the HTTP default (80) is returned as the port number.
-func GetServiceAndPort(authority string) (ServiceID, Port, error) {
+// the HTTP default (80) is returned as the port number.  If the authority
+// is a pod DNS name then the pod hostname is also returned as the 3rd return
+// value.  See https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/.
+func GetServiceAndPort(authority string, clusterDomain string) (ServiceID, Port, string, error) {
 	host, port, err := getHostAndPort(authority)
 	if err != nil {
-		return ServiceID{}, 0, err
+		return ServiceID{}, 0, "", err
 	}
 	domains := strings.Split(host, ".")
-	// S.N.svc.cluster.local
-	if len(domains) != 5 {
-		return ServiceID{}, 0, fmt.Errorf("Invalid k8s service %s", host)
+	suffix := append([]string{"svc"}, strings.Split(clusterDomain, ".")...)
+	n := len(domains)
+	// S.N.{suffix}
+	if n < 2+len(suffix) {
+		return ServiceID{}, 0, "", fmt.Errorf("Invalid k8s service %s", host)
 	}
-
-	// Needs to have `".svc."` to be a valid k8s svc
-	if domains[2] != "svc" {
-		return ServiceID{}, 0, fmt.Errorf("Invalid k8s service %s", host)
+	for i, subdomain := range domains[n-len(suffix):] {
```

Comment:
If i'm not missing anything you could rewrite this loop into:

```go
// svc.{clusterDomain} != 'host' suffix
if !strings.HasSuffix(host, "svc."+clusterDomain) {
	return ServiceID{}, 0, "", fmt.Errorf("Invalid k8s service %s", host)
}
```
which is easier to understand.